text/0047-storage.md

```diff
@@ -0,0 +1,217 @@
+---
+title: storage
+tags: RFC, draft
+---
+
+# bee-storage
++ Feature name: `storage`
++ Start date: 2020-07-15
++ RFC PR: [iotaledger/bee-rfcs#47](https://github.com/iotaledger/bee-rfcs/pull/47)
+
+# Summary
+
+An abstraction for the persistent module which will be exclusively used by the Bee crates
+
+# Motivation
+
+The Tangle is a DAG where each vertex's data is a transaction,
+this data should of course be persisted, it is better to have this logic
+and functionality grouped into one separate crate that exposes a clear API, this will also allow for implementing different database backends and ease of operation usage.
+
+The storage operation can be performed by using the methods associated with the structure, which are implemented through macros and compiled statically.
+
+This crate supports different databases, which can be indicated by using the [features](https://doc.rust-lang.org/cargo/reference/features.html#the-features-section).
+
+Note that different databases have different basic operations as well as data model definitions, which make it hard to define a general operation trait (like insertion, deletion, update, etc.) for a general structure. The target of this crate, however, is mainly for IOTA protocol, so defining operation macros for different structures used in IOTA protocal (like transaction, milestone, state_delta, etc.) is a proper approach. In this way, in addition to defining the start and shutdown methods in the `Backend` trait (which has `start` and `shutdown` methods), different databases can be used as long as the operations macros associtated to the database and a given object are implemented in the object space. Based on this design, a user can switch to a different database backend through cargo flag feature.
+
+
+use cases:
+
+1. Storing and accessing IOTA objects with databases in an efficent way
+2. Increase felixablity of adding/deleting/updating IOTA-object-related operations
+
+# Detailed Design
+
+- Supporting of each database should be switched by using the [features](https://doc.rust-lang.org/cargo/reference/features.html#the-features-section) section for ease of usage
+- Consider only RocksDB in the begining for ease of implementation and compatible with the current [IRI](https://github.com/iotaledger/iri) design
+- File hierarchy
+
+```
+    +-- access
+    |   +-- transaction.rs
+    |   +-- milestone.rs
+    |   +-- state_delta.rs
+    |   +-- tangle.rs
+    |   +-- snapshot.rs
+    |   +-- address.rs
+    |   +-- ..
+    +-- storage
+    |   +-- config.rs
+    |   +-- mod.rs
+    |   +-- rocksdb.rs
+    |   +-- ..
+```
+
+- Current Structures
+
+```rust
+/// The Storage struct will be created automatically by the assignment in feature
+/// Here we use rocks_db as an example for the storage structure
+
+#[cfg(feature = "rocks_db")]
+pub struct Storage {
+    inner: ::rocksdb::DB  // The storage backend
```

Comment:
Perhaps it should be the `inner` field that has the feature flag?

Reply:
Interesting, I agree it makes more sense.